src/data-model-providers/codegen/ClusterIntegration.h

```diff
@@ -0,0 +1,105 @@
+    ///     0-based array index from endpoint id, optional attribute fetching, feature map fetching
+    ///   - registering/unregistering ServerClusterInterfaces
+    ///   - determining how to create/destroy classes (the purpose of this delegate)
+    class Delegate
+    {
+    public:
+        virtual ~Delegate() = default;
+
+        /// Create the given registration for an endpoint
+        ///
+        /// When this is called, the caller has ensured that the endpointId was valid, that optionalAttributeBits are loaded
+        /// from ember according to supported attributes and that the feature map for the underlying cluster ID has been loaded.
+        ///
+        /// NOTE: optionalAttributeBits is intended for low id attributes since it supports attribute bits up to 31 only. It is
+        ///       intended for use with `OptionalAttributeSet` and NOT all clusters support this. Specific examples:
+        ///        - LevelControl::StartupCurrentLevel has ID 0x4000
+        ///        - OnOff attributes GlobalSceneControl, OnTime, OffWaitTime, StartupOnOff have ID >= 0x4000
+        ///        - ColorControl, Thermostat and DoorLock have many attributes with high ID as well
+        ///
+        /// Use optionalAttributes only if its usage makes sense for the cluster. In many instances it does,
+        /// however it is not a generic rule. Usage of it must be double-checked as sufficient.
+        virtual ServerClusterRegistration & CreateRegistration(EndpointId endpointId, unsigned zeroBasedArrayIndex,
+                                                               uint32_t optionalAttributeBits, uint32_t featureMap) = 0;
+
+        // Find the previously created cluster on the given path
+        virtual ServerClusterInterface & FindRegistration(unsigned zeroBasedArrayIndex) = 0;
+
+        // Destroy/free the given registration
+        virtual void DestroyRegistration(unsigned zeroBasedArrayIndex) = 0;
+    };
+
+    struct RegisterServerOptions
+    {
+        EndpointId endpointId;
+        ClusterId clusterId;
+        uint16_t fixedClusterServerEndpointCount; // data to aid in ember 0-based indexing
+        uint16_t maxEndpointCount;                // this is how many endpoints are supported by the delegates (0-based index)
+        bool fetchFeatureMap;                     // read feature map attribute from ember
+        bool fetchOptionalAttributes;             // read the enabling of the first 32 optional attributes from ember
+    };
+
+    /// Loads required data from ember and calls `CreateRegistration` once all the data
+    /// has been validated. Validation includes:
+    ///  - cluster exists on the given endpoint and a valid index could be found for it
+    ///  - feature map could be loaded (if applicable, otherwise it will be set to 0)
+    ///  - optional attributes were loaded
+    ///
+    /// The returned `CreateRegistration` value will be used to register to the codegen
+    /// data model provider registry
```

Comment:
These implementation details about "CreateRegistration" are odd and maybe go a bit too deep. Perhaps an example of "without RegisterServer " and "with RegisterServer" is better than the comment explanation.